### PR TITLE
Problem: We need an mlm client class template

### PR DESCRIPTION
**Pull request by perrettecl**
+ Adding MlmClient into project.xml
+ Generating sources and populate

Signed-off-by: Clement Perrette <clementperrette@eaton.com>

### CODE REVIEW

File: include/fty_common_mlm_client.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+         * handle the configuration of the consumption and production parts of the
+         * connection.
+         *
+         * \param pipe Pipe of the zactor_t
+         * \param endpoint Endpoint to connect to
+         * \param address Name of the established connection
+         * \param pollerTimeout Timeout of poller and rough interval betwenn tick() invocations (-1 to disable)
+         * \param connectionTimeout Timeout for connection attempt
+         */
+        MlmClient(zsock_t *pipe, const char *endpoint = NULL, const char *address = NULL, int pollerTimeout = -1, int connectionTimeout = 5000);
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: Change `const char *endpoint` and `const char *address` parameters to `const std::string &endpoint = ""` and `const std::string &address = ""`.

File: include/fty_common_mlm_client.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        /**
+         * \brief connect to broker malamute
+         * \return false to stop the agent, true otherwise.
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: `connect(...)` method doesn't return anything, incorrect comment.

File: include/fty_common_mlm_client.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+         * \brief Constructor.
+         *
+         * This will create and connect the mlm_client_t object. The agent must
+         * handle the configuration of the consumption and production parts of the
+         * connection.
+         *
+         * \param pipe Pipe of the zactor_t
+         * \param endpoint Endpoint to connect to
+         * \param address Name of the established connection
+         * \param pollerTimeout Timeout of poller and rough interval betwenn tick() invocations (-1 to disable)
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: betwenn -> between

File: include/fty_common_mlm_client.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+         */
+        MlmClient(zsock_t *pipe, const char *endpoint = NULL, const char *address = NULL, int pollerTimeout = -1, int connectionTimeout = 5000);
+        /**
+         * \brief connect to broker malamute
+         * \return false to stop the agent, true otherwise.
+         * \param endpoint Endpoint to connect to
+         * \param address Name of the established connection
+         * \param connectionTimeout Timeout for connection attempt
+         */
+        virtual void connect(const char *endpoint, const char *address, int connectionTimeout = 5000);
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: Change `const char *endpoint` and `const char *address` parameters to `const std::string &endpoint = ""` and `const std::string &address = ""`.

File: include/fty_common_mlm_client.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
+/*  =========================================================================
+    fty_common_mlm_client - C++ class to help with malamute
+
+    Copyright (C) 2014 - 2018 Eaton
```
Comment (geraldguillaume) [inline include/fty_common_mlm_client.h]: 2018-2019

File: src/fty_common_mlm_client.cc [hunk 6]
```diff
@@ -0,0 +1,145 @@
+/*  =========================================================================
+    fty_common_mlm_client - C++ class to help with malamute
+
+    Copyright (C) 2014 - 2018 Eaton
```
Comment (geraldguillaume) [inline src/fty_common_mlm_client.cc]: 2018-2019

File: include/fty_common_mlm_client.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+         * \brief Periodic callback, if enabled.
+         * \return false to stop the agent, true otherwise.
+         */
+        virtual bool tick() { return true; }
+
+        /**
+         * \brief Callback for stream messages. The callback DOESN'T take ownership of the message.
+         * \return false to stop the agent, true otherwise.
+         */
+        virtual bool handleStream(zmsg_t *message) { return true; }
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: We need to make obvious to the compiler that `msg` shan't be trifled with. Consider switching `zmsg_t *msg` to `const zmsg_t *msg` here and to other callbacks.

File: src/fty_common_mlm_client.cc [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                    if (!tick()) {
+                        break;
+                    }
+                    m_lastTick = curClock;
+                }
+            }
+
+            if (which == m_pipe) {
+                ZmsgGuard message(zmsg_recv(m_pipe));
+                if (!handlePipe(message.get())) {
```
Comment (boricj) [inline src/fty_common_mlm_client.cc]: Add `if (message == nullptr) { log_debug("interrupted"); break; }` safety net.

File: src/fty_common_mlm_client.cc [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                    }
+                }
+                else {
+                    log_warning("Unknown malamute pattern: '%s'. Message subject: '%s', sender: '%s'.",
+                            mlm_client_command(m_client), mlm_client_subject(m_client), mlm_client_sender(m_client));
+                }
+            }
+            else {
+                ZmsgGuard message(zmsg_recv(which));
+                if (!handleOther(message.get(), which)) {
```
Comment (boricj) [inline src/fty_common_mlm_client.cc]: Add `if (message == nullptr) { log_debug("interrupted"); break; }` safety net.

File: include/fty_common_mlm_client.h [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+         * \brief Create/get the zpoller_t* object for the agent mainloop.
+         * \return The mainloop poller object.
+         */
+        virtual zpoller_t* zpoller();
+
+        mlm_client_t *m_client;
+        zsock_t *m_pipe;
+        int64_t m_lastTick;
+        int m_pollerTimeout;
+        int m_connectionTimeout;
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: Consider removing `m_connectionTimeout` as its value isn't required once connection is complete.

File: include/fty_common_mlm_client.h [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+         */
+        virtual bool handleOther(zmsg_t *message, void *which) { return true; }
+
+        /**
+         * \brief Create/get the zpoller_t* object for the agent mainloop.
+         * \return The mainloop poller object.
+         */
+        virtual zpoller_t* zpoller();
+
+        mlm_client_t *m_client;
```
Comment (boricj) [inline include/fty_common_mlm_client.h]: Consider making all these properties private and add two non-virtual const getters  `mlm_client_t* client() const { return m_client; }` and `zsock_t* pipe() const { return m_pipe; }` instead.

File: src/fty_common_mlm_client.cc [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+    MlmClient::MlmClient(zsock_t *pipe, const char *endpoint, const char *address, int pollerTimeout, int connectionTimeout)
+        : m_client(mlm_client_new()),
+        m_pipe(pipe),
+        m_lastTick(zclock_mono()),
+        m_pollerTimeout(pollerTimeout),
+        m_connectionTimeout(connectionTimeout),
+        m_defaultZpoller(nullptr)
+    {
+
```
Comment (boricj) [inline src/fty_common_mlm_client.cc]: Remove useless empty line.

**Pull request merged**